ext/geos_c_impl/ruby_more.c

```diff
@@ -0,0 +1,65 @@
+/*
+	Utilities for the ruby CAPI
+*/
+
+#ifndef RGEO_GEOS_RUBY_MORE_INCLUDED
+#define RGEO_GEOS_RUBY_MORE_INCLUDED
+
+#include <ruby.h>
+
+#include "preface.h"
+#include "ruby_more.h"
+
+RGEO_BEGIN_C
+
+struct funcall_args
+{
+	VALUE recv;
+	ID mid;
+	int argc;
+	VALUE *argv;
+};
+
+static VALUE inner_funcall(VALUE args_)
+{
+	struct funcall_args *args = (struct funcall_args *)args_;
+	return rb_funcallv(
+		args->recv,
+		args->mid,
+		args->argc,
+		args->argv);
+}
```

Comment:
why do we need to break out the `inner_funcall` like this?

Reply:
`rb_protect` take a function as its first argument, when you look at the source code, it is related to how ruby handles `longjmp`: it goes like:

1. prepare security in case of longjmp
2. call the actual fn
3. remove the security
4. finally return

I could have done all that in the `rb_protect_funcall` method but it would have meant copying a lot of ruby source code, and using internal macros!

To go further you could check https://blog.peterzhu.ca/ruby-c-ext-part-8/ and the source code of `rb_protect` and `rb_ensure` for instance